File: Common/src/main/java/at/petrak/hexcasting/client/ClientRenderHelper.java
package at.petrak.hexcasting.client;

import at.petrak.hexcasting.api.casting.math.HexPattern;
import at.petrak.hexcasting.api.client.ClientCastingStack;
import at.petrak.hexcasting.api.pigment.FrozenPigment;
import at.petrak.hexcasting.client.render.RenderLib;
import at.petrak.hexcasting.xplat.IXplatAbstractions;
import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.math.Vector3f;
import net.minecraft.client.renderer.GameRenderer;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.ShaderInstance;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.Vec2;
import net.minecraft.world.phys.Vec3;

import java.util.List;

public class ClientRenderHelper {

    public static void renderCastingStack(PoseStack ps, Player player, MultiBufferSource bufferSource, float pticks) {
        for(int k = 0; k < ClientCastingStack.getPatterns().size(); k++){
            if(ClientCastingStack.patterns.get(k) == null) continue;
            ps.pushPose();
            int lifetime = ClientCastingStack.patterns.get(k).lifetime();
            float lifetimeOffset = lifetime <= 5f ? ((5f - lifetime) / 5f) : 0f;
            ps.mulPose(Vector3f.YP.rotationDegrees((float)(((player.level.getGameTime()+pticks)*(Math.sin(k*12.543565f) * 3.4f)*(k/12.43d)%360)+(1+ k)*45f)));
            ps.translate(0,1f + (Math.sin(k)*0.75f),0.75f+((Math.cos(k/8f))*0.25f) + Math.cos((player.level.getGameTime()+pticks)/(7f + (k/4f)))*0.065f);
            ps.scale(1/24f* (1-lifetimeOffset), 1/24f* (1-lifetimeOffset), 1/24f * (1-lifetimeOffset));
            ps.translate(0,Math.floor(k/8f),0);
            ps.translate(0,Math.sin((player.level.getGameTime()+pticks)/(7f + (k/8f))),0f);
            HexPattern pattern = ClientCastingStack.getPattern(k);
            if(pattern == null) {
                ps.popPose();
                continue;
            }
            ShaderInstance oldShader = RenderSystem.getShader();
            RenderSystem.setShader(GameRenderer::getPositionColorShader);
            RenderSystem.enableDepthTest();
            RenderSystem.disableCull();
            Vec2 com1 = pattern.getCenter(1);
            List<Vec2> lines1 = pattern.toLines(1, Vec2.ZERO);
            float maxDx = -1f;
            float maxDy = -1f;
            for (Vec2 line : lines1) {
                float dx = Math.abs(line.x - com1.x);
                if (dx > maxDx) {
                    maxDx = dx;
                }
                float dy = Math.abs(line.y - com1.y);
                if (dy > maxDy) {
                    maxDy = dy;
                }
            }
            float scale = Math.min(3.8f, Math.min(16/2.5f/maxDx, 16/2.5f/maxDy));
            Vec2 com2 = pattern.getCenter(scale);
            List<Vec2> lines2 = pattern.toLines(scale, com2.negated());
            for(int i = 0; i < lines2.size(); i++){
                Vec2 line = lines2.get(i);
                lines2.set(i, new Vec2(line.x, -line.y));
            }
            float variance = 0.65f;
            float speed = 0.1f;
            float stupidHash = player.hashCode();
            List<Vec2> zappy = RenderLib.makeZappy(lines2, RenderLib.findDupIndices(pattern.positions()),
                    5, variance, speed, 0.2f, 0f,
                    1f, stupidHash);
            int outer = IXplatAbstractions.INSTANCE.getPigment(player).getColorProvider().getColor(ClientTickCounter.getTotal() / 2f, Vec3.ZERO);
            int rgbOnly = outer & 0x00FFFFFF;
            int newAlpha = outer >>> 24;
            if(lifetime <= 60){
                newAlpha = (int) Math.floor((lifetime/60f) * 255);
            }
            int newARGB = (newAlpha << 24) | rgbOnly;
            int inner = RenderLib.screen(newARGB);
            RenderLib.drawLineSeq(ps.last().pose(), zappy, 0.35f, 0f, newARGB, newARGB);
            RenderLib.drawLineSeq(ps.last().pose(), zappy, 0.14f, 0.01f, inner, inner);
            ps.popPose();
            RenderSystem.setShader(() -> oldShader);
            RenderSystem.enableCull();
        }
    }
}

File: Common/src/main/java/at/petrak/hexcasting/api/client/ClientRenderHelper.kt
@file:JvmName("ClientRenderHelper")
package at.petrak.hexcasting.api.client

import at.petrak.hexcasting.client.ClientTickCounter
import at.petrak.hexcasting.client.render.drawLineSeq
import at.petrak.hexcasting.client.render.findDupIndices
import at.petrak.hexcasting.client.render.makeZappy
import at.petrak.hexcasting.client.render.screenCol
import at.petrak.hexcasting.xplat.IClientXplatAbstractions
import at.petrak.hexcasting.xplat.IXplatAbstractions
import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.math.Axis
import net.minecraft.client.renderer.GameRenderer
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.Vec2
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.sin


fun renderCastingStack(ps: PoseStack, player: Player, pticks: Float) {
    val stack = IClientXplatAbstractions.INSTANCE.getClientCastingStack(player)

    for (k in 0 until stack.getPatterns().size) {
        val patternRenderHolder = stack.getPatternHolder(k) ?: continue
        val pattern = patternRenderHolder.pattern
        val lifetime = patternRenderHolder.lifetime
        val lifetimeOffset = if (lifetime <= 5f) (5f - lifetime) / 5f else 0f

        ps.pushPose()
        ps.mulPose(Axis.YP.rotationDegrees(((player.level().gameTime + pticks) * (sin(k * 12.543565f) * 3.4f) * (k / 12.43f) % 360 + (1 + k) * 45f)))
        ps.translate(0.0, 1 + sin(k.toDouble()) * 0.75, 0.75 + cos((k / 8.0)) * 0.25 + cos((player.level().gameTime + pticks) / (7 + k / 4)) * 0.065)
        ps.scale(1 / 24f * (1 - lifetimeOffset), 1 / 24f * (1 - lifetimeOffset), 1 / 24f * (1 - lifetimeOffset))
        ps.translate(0.0, floor((k / 8.0)), 0.0)
        ps.translate(0.0, sin((player.level().gameTime + pticks) / (7.0 + k / 8.0)), 0.0)

        val oldShader = RenderSystem.getShader()
        RenderSystem.setShader { GameRenderer.getPositionColorShader() }
        RenderSystem.enableDepthTest()
        RenderSystem.disableCull()
        val com1 = pattern.getCenter(1f)
        val lines1 = pattern.toLines(1f, Vec2.ZERO)
        var maxDx = -1f
        var maxDy = -1f
        for (line in lines1) {
            val dx = abs(line.x - com1.x)
            if (dx > maxDx) {
                maxDx = dx
            }
            val dy = abs(line.y - com1.y)
            if (dy > maxDy) {
                maxDy = dy
            }
        }
        val scale = 3.8f.coerceAtMost((16 / 2.5f / maxDx).coerceAtMost(16 / 2.5f / maxDy))
        val com2 = pattern.getCenter(scale)
        val lines2 = pattern.toLines(scale, com2.negated()).toMutableList()
        for (i in lines2.indices) {
            val line = lines2[i]
            lines2[i] = Vec2(line.x, -line.y)
        }
        val variance = 0.65f
        val speed = 0.1f
        val stupidHash = player.hashCode().toDouble()
        val zappy: List<Vec2> = makeZappy(lines2, findDupIndices(pattern.positions()),
                5, variance, speed, 0.2f, 0f,
                1f, stupidHash)
        val outer: Int = IXplatAbstractions.INSTANCE.getPigment(player).colorProvider.getColor(
                ClientTickCounter.getTotal() / 2f,
                patternRenderHolder.getColourPos(player.random))
        val rgbOnly = outer and 0x00FFFFFF
        var newAlpha = outer ushr 24
        if (lifetime <= 60) {
            newAlpha = floor((lifetime / 60f * 255).toDouble()).toInt()
        }
        val newARGB = newAlpha shl 24 or rgbOnly
        val inner: Int = screenCol(newARGB)
        drawLineSeq(ps.last().pose(), zappy, 0.35f, 0f, newARGB, newARGB)
        drawLineSeq(ps.last().pose(), zappy, 0.14f, 0.01f, inner, inner)
        ps.popPose()
        RenderSystem.setShader { oldShader }
        RenderSystem.enableCull()
    }
}